File: job-2022-autumn/19.remove-nth-node-from-end-of-list.js
/*
 * @lc app=leetcode id=19 lang=javascript
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
    const fakeHead = new ListNode();
    fakeHead.next = head;

    let prevNode = fakeHead;
    let nextNode = fakeHead;

    for (let i = 0; i < n; i++) {
        nextNode = nextNode.next;
    }

    while (nextNode.next !== null) {
        prevNode = prevNode.next;
        nextNode = nextNode.next;
    }

    if (prevNode.next !== null) {
        prevNode.next = prevNode.next.next;
    }

    return fakeHead.next;
};
// @lc code=end

File: Daily/19.remove-nth-node-from-end-of-list.cpp
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};
// @lc code=start
class Solution {
 public:
  ListNode* removeNthFromEnd(ListNode* head, int n) {
    ListNode fakeHead;
    fakeHead.next = head;

    ListNode* before = &fakeHead;
    ListNode* after = &fakeHead;

    // After is at before + n position, so before is the previous node of
    // removed one
    for (int i = 0; i < n; i++) {
      after = after->next;
    }

    while (after->next != nullptr) {
      before = before->next;
      after = after->next;
    }

    ListNode* prevNode = before;
    ListNode* nextNode = before->next->next;

    prevNode->next = nextNode;

    return fakeHead.next;
  }
};
// @lc code=end
